Source/Core/AudioCommon/OpenALStream.cpp

```diff
@@ -178,13 +178,34 @@ void OpenALStream::SoundLoop()
   surround_capable = false;
 #endif
 
-  u32 ulFrequency = m_mixer->GetSampleRate();
-  numBuffers = SConfig::GetInstance().iLatency + 2;  // OpenAL requires a minimum of two buffers
+  u32 ul_frequency = m_mixer->GetSampleRate();
+  // OpenAL requires a minimum of two buffers
+  num_buffers = SConfig::GetInstance().iNumBuffers;
 
-  realtimeBuffer.reserve(OAL_MAX_SAMPLES * STEREO_CHANNELS);
-  sampleBuffer.reserve(OAL_MAX_SAMPLES * SURROUND_CHANNELS * OAL_MAX_BUFFERS);
-  uiBuffers.reserve(OAL_MAX_BUFFERS);
-  uiSource = 0;
+  if (num_buffers < 2)
+    num_buffers = 2;
+
+  // calculate latency (samples) per buffer
+  uint samples_per_buffer;
+  if (SConfig::GetInstance().iLatency > 0)
+  {
+	  samples_per_buffer = ul_frequency / 1000 * SConfig::GetInstance().iLatency / num_buffers;
+  }
+  else
+  {
+	  samples_per_buffer = ul_frequency / 1000 * 1 / num_buffers;
+  }
+
+  // DPL2 needs a minimum number of samples to work (FWRDURATION)
+  if (surround_capable && samples_per_buffer < 240)
+  {
+	  samples_per_buffer = 240;
+  }
+
+  realtime_buffer.reserve(samples_per_buffer * STEREO_CHANNELS);
+  sample_buffer.reserve(samples_per_buffer * STEREO_CHANNELS * 20);
```

Comment:
I guess I should comment about it on the code.
sample_buffer is the buffer for SoundTouch, which can stretch the number of samples up to ten times (as we set the minimum rate to 0.1). I am multiplying by 20 just to be safe.